File: SwpTextView/Object-C/Code/SwpTextView.m
//
//  SwpTextView.m
//  swp_song
//
//  Created by swp_song on 2018/1/18.
//  Copyright © 2018年 swp_song. All rights reserved.
//

#import "SwpTextView.h"

/* ---------------------- Tool       ---------------------- */
#import "SwpTextViewUtils.h"
/* ---------------------- Tool       ---------------------- */

static CGFloat const kSwpTextViewAcquiesceFontSize = 15.0f;

@interface SwpTextView ()

#pragma mark - UI   Propertys
/* ---------------------- UI   Property ---------------------- */
/* 显示 placeholder View */
@property (nonatomic, strong) UILabel *placeholderView;
/* ---------------------- UI   Property ---------------------- */
#pragma mark - Data Propertys
/* ---------------------- Data Property ---------------------- */
/* 动画时长 */
@property (nonatomic, assign) NSTimeInterval aHiddenAnimationTime;

/* SwpTextView 回调方法，用户输入文字变化调用 */
@property (nonatomic, copy, setter = swp_textViewChangeEvent:) SwpTextViewTextChangeBlock swp_textViewChangeEvent;
/* ---------------------- Data Property ---------------------- */

@end

@implementation SwpTextView

/**
 *  @author swp_song
 *
 *  @brief  initWithFrame:textContainer:    ( Override Init )
 *
 *  @param  frame           frame
 *
 *  @param  textContainer   textContainer
 *
 *  @return UITextView
 */
- (instancetype)initWithFrame:(CGRect)frame textContainer:(NSTextContainer *)textContainer {
    if (self = [super initWithFrame:frame textContainer:textContainer]) {
        [self config];
    }
    return self;
}

/**
 *  @author swp_song
 *
 *  @brief  initWithCoder:  ( Override Init )
 *
 *  @param  aDecoder    aDecoder
 *
 *  @return UITextView
 */
- (instancetype)initWithCoder:(NSCoder *)aDecoder {
    if (self = [super initWithCoder:aDecoder]) {
        [self config];
    }
    return self;
}

/**
 *  @author swp_song
 *
 *  @brief  config  ( config )
 */
- (void)config {
    [self setProperty];
    [self setUpUI];
}

/**
 *  @author swp_song
 *
 *  @brief  layoutSubviews  ( Override layoutSubviews )
 */
- (void)layoutSubviews {
    [super layoutSubviews];
    [self setUIFrame];
}

/**
 *  @author swp_song
 *
 *  @brief  dealloc (  )
 */
- (void)dealloc {
    
   
    
#if DEBUG
    NSLog(@"%@ - %s - dealloc", self.class, __FUNCTION__);
#else
    
#endif
    
    // 移除通知
    [[NSNotificationCenter defaultCenter] removeObserver:self];
}

/*!
 *  @author swp_song
 *
 *  @brief  setupUI ( 添加控件 )
 */
- (void)setUpUI {
    [self addSubview:self.placeholderView];
}

/**
 *  @author swp_song
 *
 *  @brief  setProperty ( 设置属性 )
 */
- (void)setProperty {
    self.aHiddenAnimationTime = 0.5;
    self.font = [UIFont systemFontOfSize:kSwpTextViewAcquiesceFontSize];
    [self aPlaceholder:self.swp_Text == nil ? @"" : self.swp_Text animate:self.aHiddenAnimationTime];
    [[NSNotificationCenter defaultCenter] addObserver:self selector:@selector(textDidChange:) name:UITextViewTextDidChangeNotification object:self];
}

/**
 *  @author swp_song
 *
 *  @brief  setUIFrame  ( 设置 UI 控件的 frame )
 */
- (void)setUIFrame {
    CGFloat placeholderX = 8.0;
    self.placeholderView.frame = CGRectMake(placeholderX, 3, self.frame.size.width - placeholderX * 2.0, 25);
}

/**
 *  @author swp_song
 *
 *  @brief  textDidChange   ( 通知绑定方法 )
 *
 *  @param  notification    notification
 */
- (void)textDidChange:(NSNotification *)notification {
    _swp_Text = super.text;

    if (self.swp_textViewChangeEvent) self.swp_textViewChangeEvent(self, super.text);

    if ([self.swp_Delegate  respondsToSelector:@selector(swp_textView:changeText:)]) {
        [self.swp_Delegate swp_textView:self changeText:super.text];
    }

    [self aPlaceholder:super.text animate:self.aHiddenAnimationTime];
}

/**
 *  @author swp_song
 *
 *  @brief  showPlaceholder:animateDuration:    ( 判断 Placeholder 显示，隐藏 )
 *
 *  @param  text    text
 */
- (void)aPlaceholder:(NSString *)text animate:(NSTimeInterval)animate {
    //  Textview 长度为0
    if (self.text.length == 0) {
        
        //  判断是否为删除键
        if ([text isEqualToString:@""]) {
            [self swp_placeholderHidden:NO duration:animate];
        } else {
            [self swp_placeholderHidden:YES duration:animate];
        }
        
    } else {
        // Textview 长度不为 0
        if (self.text.length == 1) {
            //  Textview 长度为1时候
            if ([text isEqualToString:@""]) {
                //  判断是否为删除键
                [self swp_placeholderHidden:NO duration:animate];
                
            } else {
                //  不是删除
               [self swp_placeholderHidden:YES duration:animate];
            }
        } else {
            //  长度不为1时候
            [self swp_placeholderHidden:YES duration:animate];
            
        }
    }
}

#pragma mark - Public Methods

/**
 *  @author swp_song
 *
 *  @brief  swp_info ( SwpTextView 信息 )
 *
 *  @return NSDictionary
 */
+ (NSDictionary *)swp_info {
    return [SwpTextViewUtils swpTextViewUtilsGetInformation];
}

/**
 *  @author swp_song
 *
 *  @brief  swp_version ( SwpTextView 版本信息 )
 *
 *  @return NSString
 */
+ (NSString *)swp_version {
    return self.class.swp_info[@"Version"];
}

/**
 *  @author swp_song
 *
 *  @brief  swp_init ( 快速初始化 )
 *
 *  @return SwpTextView
 */
+ (instancetype)swp_textView {
    return [self.class new];
}

/**
 *  @author swp_song
 *
 *  @brief  swp_init ( 快速初始化 )
 */
+ (__kindof SwpTextView * (^)(void))swp_init {
    return ^(void) {
        return [self.class swp_textView];
    };
}


/**
 *  @author swp_song
 *
 *  @brief  swpText ( 设置显示文本 )
 */
- (__kindof SwpTextView * _Nonnull (^)(NSString * _Nonnull))swp_text {
    return ^(NSString *text) {
        [self swp_placeholderHidden:YES duration:0];
        self.text = text;
        return self;
    };
}

/**
 *  @author swp_song
 *
 *  @brief  swp_placeholderHidden:duration: ( 隐藏，显示 Placeholder  )
 *
 *  @param  isDisplay   isDisplay
 *
 *  @param  duration    duration
 */
- (void)swp_placeholderHidden:(BOOL)isHidden duration:(NSTimeInterval)duration {
    [UIView animateWithDuration:duration animations:^{
        self.placeholderView.alpha = !isHidden;
    }];
}

/**
 *  @author swp_song
 *
 *  @brief  swp_delegate ( 设置代理 )
 */
- (__kindof SwpTextView * _Nonnull (^)(id<SwpTextViewDelegate> _Nonnull))swp_delegate {
    
    return ^(id<SwpTextViewDelegate>delegate) {
        if (!delegate) return self;
        self.swp_Delegate = delegate;
        return self;
    };
}

/**
 *  @author swp_song
 *
 *  @brief  swp_systemFontSize  ( 设置，输入文本字体大小，系统字体 )
 */
- (__kindof SwpTextView * _Nonnull (^)(CGFloat))swp_systemFontSize {
    
    return ^(CGFloat size) {
        self.font = [UIFont systemFontOfSize:size];
        return self;
    };
}

/**
 *  @author swp_song
 *
 *  @brief  swp_textColor  ( 设置，输入文本字体颜色 )
 */
- (__kindof SwpTextView * _Nonnull (^)(UIColor * _Nonnull))swp_textColor {
    
    return ^(UIColor *color) {
        self.textColor = color;
        return self;
    };
}

/**
 *  @author swp_song
 *
 *  @brief  swp_font  ( 设置，输入文本字体 )
 */
- (__kindof SwpTextView * _Nonnull (^)(UIFont * _Nonnull))swp_font {
    
    return ^(UIFont *font) {
        self.font = font;
        return self;
    };
}


/**
 *  @author swp_song
 *
 *  @brief  swp_placeholder ( 设置 Placeholder )
 */
- (__kindof SwpTextView * _Nonnull (^)(NSString * _Nonnull))swp_placeholder {
    return ^(NSString *placeholder) {
        self.placeholderView.text = placeholder;
        return self;
    };
}

/**
 *  @author swp_song
 *
 *  @brief  swp_placeholderColor  ( 设置，placeholder 字体颜色 )
 */
- (__kindof SwpTextView * _Nonnull (^)(UIColor * _Nonnull))swp_placeholderColor {
 
    return ^(UIColor *color) {
        self.placeholderView.textColor = color;
        return self;
    };
}

/**
 *  @author swp_song
 *
 *  @brief  swp_placeholderFont ( 设置，placeholder 字体 )
 */
- (__kindof SwpTextView * _Nonnull (^)(UIFont * _Nonnull))swp_placeholderFont {
    return ^(UIFont *font) {
        self.placeholderView.font = font;
        return self;
    };
}

/**
 *  @author swp_song
 *
 *  @brief  swp_keyboardType    ( 设置，键盘样式 )
 */
- (__kindof SwpTextView * _Nonnull (^)(UIKeyboardType))swp_keyboardType {
    
    return ^(UIKeyboardType keyboardType) {
        self.keyboardType = keyboardType;
        return self;
    };
}

/**
 *  @author swp_song
 *
 *  @brief  placeholderHiddenAnimationTime  ( 设置，placeholder 隐藏动画时长 )
 */
- (__kindof SwpTextView * _Nonnull (^)(NSTimeInterval))placeholderHiddenAnimationTime {
    return ^(NSTimeInterval time) {
        self.aHiddenAnimationTime = time;
        return self;
    };
}

/**
 *  @author swp_song
 *
 *  @brief  borderWidth ( 设置，显示边框宽度 )
 */
- (__kindof SwpTextView * _Nonnull (^)(CGFloat))swp_borderWidth {
    
    return ^(CGFloat borderWidth) {
        if (!borderWidth) return self;
        self.layer.borderWidth = borderWidth;
        return self;
    };
}

/**
 *  @author swp_song
 *
 *  @brief  cornerRadius ( 设置，显示边框圆角弧度 )
 */
- (__kindof SwpTextView * _Nonnull (^)(CGFloat))swp_cornerRadius {
    return ^(CGFloat cornerRadius) {
        if (!cornerRadius) return self;
        self.layer.cornerRadius     = cornerRadius;
        self.layer.masksToBounds    = YES;
        return self;
    };
}

/**
 *  @author swp_song
 *
 *  @brief  swp_borderColor ( 设置，显示边框颜色 )
 */
- (__kindof SwpTextView * _Nonnull (^)(UIColor * _Nonnull))swp_borderColor {
    return ^(UIColor *borderColor) {
        if (!borderColor) return self;
        self.layer.borderColor = borderColor.CGColor;
        return self;
    };
}

/**
 *  @author swp_song
 *
 *  @brief  swp_backgroundColor ( 设置，显示背景颜色 )
 */
- (__kindof SwpTextView * _Nonnull (^)(UIColor * _Nonnull))swp_backgroundColor {
    return ^(UIColor *backgroundColor) {
        if (!backgroundColor) return self;
        self.backgroundColor = backgroundColor;
        return self;
    };
}

/**
 *  @author swp_song
 *
 *  @brief  swp_textViewChangeEvent:  ( SwpTextView 回调方法，用户输入文字变化调用 )
 *
 *  @param  changeEvent changeEvent
 */
- (void)swp_textViewChangeEvent:(SwpTextViewTextChangeBlock)changeEvent {
    _swp_textViewChangeEvent = changeEvent;
}

/**
 *  @author swp_song
 *
 *  @brief  swp_textViewChangeEventChain:   ( SwpTextView 回调方法，用户输入文字变化调用 )
 */
- (__kindof SwpTextView * _Nonnull (^)(SwpTextViewTextChangeBlock _Nonnull))swp_textViewChangeEventChain {
    
    return ^(SwpTextViewTextChangeBlock changeEvent) {
        [self swp_textViewChangeEvent:changeEvent];
        return self;
    };
}

#pragma mark - Init UI Methods
- (UILabel *)placeholderView {
    
    return !_placeholderView ? _placeholderView = ({
        UILabel *label = [UILabel new];
        label.text      = @"placeholder";
        label.font      = [UIFont systemFontOfSize:kSwpTextViewAcquiesceFontSize];
        label.textColor = [UIColor lightGrayColor];
        label;
    }) : _placeholderView;
}

/*
// Only override drawRect: if you perform custom drawing.
// An empty implementation adversely affects performance during animation.
- (void)drawRect:(CGRect)rect {
    // Drawing code
}
*/

@end

File: SwpTextView/Swift/Code/SwpTextView.swift
//
//  SwpTextView.swift
//  swp_song
//
//  Created by swp-song on 2019/1/20.
//  Copyright © 2019 swp-song. All rights reserved.
//

import UIKit

open class SwpTextView: UITextView {
    
    // MARK: - Public Property
    public typealias SwpTextViewDidChangeEvent = (_ textView : SwpTextView, _ text : String) -> Void
    
    public var swp_text : String {
        return super.text
    }
    
    public weak var swp_delegate : SwpTextViewDelegate? = nil;
    public var swp_textViewDidChangeEvent : SwpTextViewDidChangeEvent? = nil
    
    // MARK: - Private Property
    private let aHiddenAnimationTime : TimeInterval   = 0.5
    private let kPlaceholderDefaultFontSize : CGFloat = 15.0
    
    private lazy var placeholderView : UILabel = {
        let label : UILabel = UILabel()
        label.text      = "placeholder"
        label.font      = UIFont.systemFont(ofSize: kPlaceholderDefaultFontSize)
        label.textColor = .lightGray
        return label;
    }()
    
    // MARK: - Override Function
    override public init(frame: CGRect, textContainer: NSTextContainer?) {
        super.init(frame: frame, textContainer: textContainer)
        self.config()
    }
    
    required public init?(coder aDecoder: NSCoder) {
        super.init(coder: aDecoder)
        self.config()
    }
    
    override open func layoutSubviews() {
        super.layoutSubviews()
        self.setUIFrame()
    }
    
    deinit {
        #if DEBUG
        print("Class = \(SwpTextView.self), \(#function)")
        #else
        #endif
        NotificationCenter.default.removeObserver(self, name: UITextView.textDidChangeNotification, object: self)
    }
    
    /*
    // Only override draw() if you perform custom drawing.
    // An empty implementation adversely affects performance during animation.
    override func draw(_ rect: CGRect) {
        // Drawing code
    }
    */

}

// MARK: - Public SwpTextView Function
public extension SwpTextView {
    
    ///
    /// # set placeholder hidden
    ///
    /// - Parameters:
    ///   - isHidden:    isHidden
    ///   - animateTime: animateTime
    /// - Returns: SwpTextView
    @discardableResult public func swp_placeholderHidden(_ isHidden : Bool, animateTime : TimeInterval = 0.5) -> Self {
        self.isPlaceholderHidden(isHidden, animate: animateTime)
        return self
    }
    
    ///
    /// # set text
    ///
    /// - Parameter text: text
    /// - Returns: SwpTextView
    @discardableResult public func swp_text(_ text : String) -> Self {
        self.swp_placeholderHidden(true, animateTime: 0)
        self.text = text
        return self
    }
    
    
    ///
    /// set delegate
    ///
    /// - Parameter delete: delete
    /// - Returns: SwpTextView
    @discardableResult public func swp_delegate(_ delete : SwpTextViewDelegate?) -> Self {
        self.swp_delegate = delete
        return self
    }
    
    ///
    /// # did change block
    ///
    /// - Parameter didChangeEvent: didChangeEvent
    /// - Returns: SwpTextView
    @discardableResult public func swp_textViewDidChangeEvent(_ didChangeEvent : @escaping SwpTextViewDidChangeEvent) -> Self {
        self.swp_textViewDidChangeEvent = didChangeEvent
        return self
    }
}

// MARK: - Public Placeholder Function
extension SwpTextView {
    
    
    ///
    /// # set placeholder text
    ///
    /// - Parameter placeholder: placeholder
    /// - Returns: SwpTextView
    @discardableResult public func swp_placeholder(_ placeholder : String?) -> Self {
        self.placeholderView.text = placeholder
        return self
    }
    
    ///
    /// # set placeholder text font
    ///
    /// - Parameter font: font
    /// - Returns: SwpTextView
    @discardableResult public func swp_placeholderFont(_ font : UIFont) -> Self {
        self.placeholderView.font = font
        return self
    }
    
    ///
    /// # set placeholder text color
    ///
    /// - Parameter color: color
    /// - Returns: SwpTextView
    @discardableResult public func swp_placeholderColor(_ color : UIColor) -> Self {
        self.placeholderView.textColor = color
        return self
    }
}

// MARK: - Public UITextView Function
extension SwpTextView {
    
    ///
    /// # set backgroundColor
    ///
    /// - Parameter backgroundColor: backgroundColor
    /// - Returns: SwpTextView
    @discardableResult public func swp_backgroundColor(_ backgroundColor : UIColor?) -> Self {
        self.backgroundColor = backgroundColor;
        return self
    }
    
    ///
    /// # set text syestem font size
    ///
    /// - Parameter size: size
    /// - Returns: SwpTextView
    @discardableResult public func swp_systemFontSize(_ size : CGFloat = 15) -> Self {
        self.font = UIFont.systemFont(ofSize: size)
        return self
    }
    
    ///
    /// # set text font
    ///
    /// - Parameter font: font
    /// - Returns: SwpTextView
    @discardableResult public func swp_font(_ font : UIFont?) -> Self {
        self.font = font
        return self
    }
    
    ///
    /// # set text color
    ///
    /// - Parameter color: color
    /// - Returns: SwpTextView
    @discardableResult public func swp_textColor(_ color : UIColor?) -> Self {
        self.textColor = color
        return self
    }
    
    ///
    /// # set text keyboard type
    ///
    /// - Parameter keyboardType: keyboardType
    /// - Returns: SwpTextView
    @discardableResult public func swp_keyboardType(_ keyboardType : UIKeyboardType) -> Self {
        self.keyboardType = keyboardType
        return self
    }
}

// MARK: - Public Layer Function
extension SwpTextView {
    
    
    ///
    /// # set borderWidth
    ///
    /// - Parameter borderWidth: borderWidth
    /// - Returns: SwpTextView
    @discardableResult public func swp_borderWidth(_ borderWidth : CGFloat) -> Self {
        self.layer.borderWidth = borderWidth
        return self
    }
    
    ///
    /// # set cornerRadius
    ///
    /// - Parameter cornerRadius: cornerRadius
    /// - Returns: SwpTextView
    @discardableResult public func swp_cornerRadius(_ cornerRadius : CGFloat) -> Self {
        self.layer.cornerRadius = cornerRadius
        self.layer.masksToBounds = true
        return self
    }
    
    ///
    /// # set borderColor
    ///
    /// - Parameter borderColor: borderColor
    /// - Returns: SwpTextView
    @discardableResult public func swp_borderColor(_ borderColor : UIColor) -> Self {
        self.layer.borderColor = borderColor.cgColor
        return self
    }
}


// MARK: - Private Set UI
private extension SwpTextView {
    
    private func config() -> Void {
        self.setUpUI()
        self.propertyConfig()
    }
    
    private func setUpUI() -> Void {
        self.addSubview(self.placeholderView)
    }
    
    private func setUIFrame() -> Void {
        let placeholderX : CGFloat = 8.0
        self.placeholderView.frame = CGRect(x: placeholderX, y: 3.0, width: self.frame.size.width - placeholderX * 2.0, height: 25.0)
    }
    
    private func propertyConfig() -> Void {
        self.font = UIFont.systemFont(ofSize: kPlaceholderDefaultFontSize)
        NotificationCenter.default.addObserver(self, selector: #selector(textDidChange), name: UITextView.textDidChangeNotification, object: self)
    }
    
    @objc private func textDidChange(_ notification : Notification) -> Void {
        let text : String = super.text
        self.swp_textViewDidChangeEvent?(self, text)
        self.swp_delegate?.swp_textViewDidChange(self, text: text)
        self.aPlaceholderHidden(text, animate: self.aHiddenAnimationTime)
    }
}


// MARK: - Placeholder Hidden & Dispaly Fuction
private extension SwpTextView {

    private func aPlaceholderHidden(_ string : String, animate: TimeInterval) -> Void {
        if self.text.count == 0 {
            if string == "" {
                self.isPlaceholderHidden(false, animate: animate)
            } else {
                self.isPlaceholderHidden(true, animate: animate)
            }
        } else {
            
            if self.text.count == 1 {
                if self.text == "" {
                    self.isPlaceholderHidden(false, animate: animate)
                } else {
                    self.isPlaceholderHidden(true, animate: animate)
                }
            } else {
                self.isPlaceholderHidden(true, animate: animate)
            }
        }
    }
    
    private func isPlaceholderHidden(_ isHidden : Bool, animate: TimeInterval) {
        UIView.animate(withDuration: animate) { //[weak self] in
            self.placeholderView.alpha = isHidden ? 0 : 1
        };
    }
}

